MAYA-127045 Add reparenting rules for Shaders, NodeGraphs and Materials.

frohnej-adsk opened the pull request:
MAYA-127045 Add reparenting rules for UsdShadeShader, UsdShadeNodeGraph and UsdShadeMaterial.

Users are able to reparent USD items in the Outliner. However, in some cases, reparenting does not make sense and can lead to problematic situations in LookdevX. The following rules are established, when a user tries to reparent items in the Outliner:

- UsdShadeShader can only have UsdShadeNodeGraph/UsdShadeMaterial as parent.
- UsdShadeNodeGraph can only have a UsdShadeNodeGraph/UsdShadeMaterial as parent.
- UsdShadeMaterial cannot have UsdShadeShader/UsdShadeNodeGraph/UsdShadeMaterial as parent.

If one of the rules is broken, reparenting is prevented.

JGamache-autodesk commented:
You are actually implementing rules that are defined in USD.
The documentation for NodeGraph is [here](https://github.com/PixarAnimationStudios/USD/blob/dev/pxr/usd/usdShade/schema.usda#L55). The reason for this encapsulation is documented [here](https://github.com/PixarAnimationStudios/USD/blob/dev/pxr/usd/usdShade/schema.usda#L123).

seando-adsk reviewed (changes requested):
This requires a unit test, similar to this one that tests the existing gprim restriction:
https://github.com/Autodesk/maya-usd/blob/dev/test/lib/ufe/testParentCmd.py#L1045

frohnej-adsk commented:
@seando-adsk I changed one of the existing tests that was unparenting a Shader, which is no longer allowed. I also added tests for the new reparenting rules.

I made the new tests only run for Maya2023 and later, because reparenting something to a Material in Maya2022 results in a runtime error for some reason: 
```
RuntimeError: Object [path/to/material] does not support parenting and therefore cannot be a parent.
```

If you approve of my changes, the PR is ready for merge.

frohnej-adsk commented on test/lib/ufe/testParentCmd.py:
```diff
@@ -997,43 +997,40 @@ def testUnparentMultiStage(self):
-            # Use relative parenting, else trying to keep absolute world
-            # position of Lambert node fails (of course).
-            cmds.parent(cubePathStr1, lambertPathStr2, w=True, r=True)
+            cmds.parent(stage1ItemPath, stage2ItemPath, w=True)
```
Previously, this call was unparenting a Shader. Unparenting or reparenting a Shader seems to fail in Maya2022 for some reason unless `relative=ture`. I think that's the only reason why the `relative` flag was set in the first place.

Since we are not unparenting a Shader anymore, the `relative` flag in is no longer necessary. If we leave it in, I feel like a reader might wonder, why it is set, so I felt like removing it would make the code more readable.

seando-adsk replied:
Okay that explains it. Thanks.
